### PR TITLE
Adding TPCdEdxCalibrationSplines class to correct the dEdx for geometrical and track topological effects

### PR DESCRIPTION
**Pull request by matthias-kleiner**
The splines which will be used to correct the dE/dx for detector effects like different pad sizes and track topologies (local dip angle: z angle - dz/dx and diffusion from electron drift) are created by the Random Forest ML algorithm. The training data is created by the O2 simulation with artificial tracks (no Landau distribution, constant dE/dx, fixed GEM amplification of 2000). For each region and for qMax/qTot an individual spline is created which leads to a total of 20 splines. 
It is planned to make further topological corrections also with the splines: relative pad and time position of a track, local inclination angle: y angle = dy/dx (improves dE/dx at low momenta) and threshold effects.

### CODE REVIEW

**Comment by davidrohr** [comment]
Could you please remove the commit "GPU: Compute timebin in TPC ZS Decoding only from current RDH and fir…" from the PR, that seems to be a rebasing artifact. The remaining change of the shift constant from 0 to 7 was a test in my local branch from which you pulled the propagation of the splines. This should not go into O2.

**Comment by davidrohr** [comment]
And please squash "fix splines class to compile" into "Adding class to store the calibration splines for dEdx corrections" to avoid unnecessary fixup commits.

**Comment by davidrohr** [comment]
If I run this with a default-constructor initialized splines object, will it then reproduce the previous dE/dx results?
How can I obtain a dEdx splines file?
Is the processing full implemented in the fillCluster? It looks like it. I would like to check what is the performance impact on the GPU.

**Comment by matthias-kleiner** [comment]
> Could you please remove the commit "GPU: Compute timebin in TPC ZS Decoding only from current RDH and fir…" from the PR, that seems to be a rebasing artifact. The remaining change of the shift constant from 0 to 7 was a test in my local branch from which you pulled the propagation of the splines. This should not go into O2.

Thanks for the comment. I will remove it then

**Comment by matthias-kleiner** [comment]
> If I run this with a default-constructor initialized splines object, will it then reproduce the previous dE/dx results?
> How can I obtain a dEdx splines file?
> Is the processing full implemented in the fillCluster? It looks like it. I would like to check what is the performance impact on the GPU.

When you specify no spline object the dE/dx result is not correct as all values will be divided by 0. I have to fix this by getting a default value of 1 from the splines or dont do the evaluation of the splines if no splines were set.  

I can send you the root file with the splines or is there some way I can store it online so everyone can access it?

The method is full implemented in fillCluster and should work if the splines were set.

**Comment by davidrohr** [comment]
Thx, it would be good if you could send me a root file for testing now.

I think the default constructor shouldn't create an all-zero spline but one that basically applies no calibration at all such that one gets the old results back. Is that possible?

**Comment by matthias-kleiner** [comment]
> Thx, it would be good if you could send me a root file for testing now. I think the default constructor shouldn't create an all-zero spline but one that basically applies no calibration at all such that one gets the old results back. Is that possible?

I think it would be possible by creating some default spline which just returns 1. I have to check this

File: GPU/GPUTracking/dEdx/GPUdEdx.h [hunk 1]
```diff
@@ -93,10 +94,26 @@ GPUdi() void GPUdEdx::fillCluster(float qtot, float qmax, int padRow, float trac
   if (snp2 > GPUCA_MAX_SIN_PHI_LOW) {
     snp2 = GPUCA_MAX_SIN_PHI_LOW;
   }
-  float factor = CAMath::Sqrt((1 - snp2) / (1 + trackTgl * trackTgl));
+  float tgl2 = trackTgl * trackTgl;
```
Comment (wiechula) [inline GPU/GPUTracking/dEdx/GPUdEdx.h:97]: `const float`

File: GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+#ifndef TPCdEdxCalibrationSplines_H
+#define TPCdEdxCalibrationSplines_H
+
+#include "FlatObject.h"
+#include "Spline2D.h"
+
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE) // code invisible on GPU and in the standalone compilation
+#include "Rtypes.h"                                       // for ClassDefNV
+#include <boost/format.hpp>
```
Comment (wiechula) [inline GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h:24]: you could consider switching to `<fmt/format.h>`

File: GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+#if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
+  /// write a class object to the file
+  int writeToFile(TFile& outf, const char* name);
+
+  /// read a class object from the file
+  static TPCdEdxCalibrationSplines* readFromFile(TFile& inpf, const char* name);
+#endif
+
+ private:
+  constexpr static unsigned int mFSplines = 10; ///< number of splines stored for each type
```
Comment (wiechula) [inline GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h:127]: I think constexpr should not start with `m`, but either `s` or simply with a capital letter. See:
https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Variable_Names#Variable_Names

File: GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+inline void TPCdEdxCalibrationSplines::setSplinesFromFile(TFile& inpf)
+{
+  FlatObject::startConstruction();
+
+  int buffSize = 0;
+  int offsets1[mFSplines];
+  int offsets2[mFSplines];
+
+  for (unsigned int ireg = 0; ireg < mFSplines; ++ireg) {
+    std::string splinename = (boost::format("spline_qMax_region%d") % ireg).str();
```
Comment (wiechula) [inline GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h:145]: consider
`std::string splinename = fmt::format("spline_qMax_region{}", ireg);`

File: GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    o2::gpu::Spline2D<float, 1>* splineTmpqMax = o2::gpu::Spline2D<float, 1>::readFromFile(inpf, splinename.data());
+    mCalibSplinesqMax[ireg] = *splineTmpqMax;
+    buffSize = alignSize(buffSize, mCalibSplinesqMax[ireg].getBufferAlignmentBytes());
+    offsets1[ireg] = buffSize;
+    buffSize += mCalibSplinesqMax[ireg].getFlatBufferSize();
+  }
+
+  for (unsigned int ireg = 0; ireg < mFSplines; ++ireg) {
+    std::string splinename = (boost::format("spline_qTot_region%d") % ireg).str();
```
Comment (wiechula) [inline GPU/GPUTracking/dEdx/TPCdEdxCalibrationSplines.h:154]: same as above, you could consider `fmt::format`

**Comment by matthias-kleiner** [comment]
> Look nice! Just minor comments.
> As David wrote, default initialization such that 1 is returned instead of 0 would be good in this case.

Thanks for the comments. The default value is now changed to 1. So the dEdx will not change if no splines are set.

**Comment by davidrohr** [comment]
I didn't have time yet to test this. But I have your root file with the splines and will try asap.

**Comment by davidrohr** [comment]
@matthias-kleiner : I finally find some time to look at this. Comparing the dE/dx from before (without splines) and after (with default splines), I see that dEdx for qTot remains the same but dEdx for qMax is very different, usually around a factor 2 smaller. Is this expected or is something wrong?

**Comment by wiechula** [comment]
@davidrohr ,
the calibration should be done in a way that the dE/dx has the same mean for the qMax and qTot estimator. Without calibration qTot should be much larger, of course.
Can you confirm that the mean of the dE/dx is the same now?

**Comment by matthias-kleiner** [comment]
> @matthias-kleiner : I finally find some time to look at this. Comparing the dE/dx from before (without splines) and after (with default splines), I see that dEdx for qTot remains the same but dEdx for qMax is very different, usually around a factor 2 smaller. Is this expected or is something wrong?

hi, when you use the old version without the splines, qMax was normalized to the padwidth "qmax *= factor / param.tpcGeometry.PadWidth(padRow);". I removed this as the normalization to the pad geometry is applied with the splines. If you want to get the same results with the default splines you should add this normalization back in.

**Comment by davidrohr** [comment]
With the file you provided, it runs correctly on CPU and GPU. (Although I cannot judge the dEdx output).
Performance penalty (dEdx with splines v.s. dEdx uncalibrated) is ~2.5% on CPU (compared to full track fit) and 3.5-5% on GPU.
I would say this is acceptable. Still, at some point one can have a look why the GPU loses more, but OK for now.

**Comment by davidrohr** [comment]
> hi, when you use the old version without the splines, qMax was normalized to the padwidth "qmax *= factor / param.tpcGeometry.PadWidth(padRow);". I removed this as the normalization to the pad geometry is applied with the splines.

OK, I understand. That makes sense. No need to change it. Was just wondering.

**Comment by davidrohr** [comment]
@matthias-kleiner : Can you apply the attached patch, it makes it compile in my standalone benchmark, just a minor change. I am wondering what is the problem with the failing ctest. On my setup, the ctest works. I'll try to have a look later.

[splines.patch.tar.gz](https://github.com/AliceO2Group/AliceO2/files/4585755/splines.patch.tar.gz)

**Comment by davidrohr** [comment]
@matthias-kleiner : Please add also the attached patch, it fixes the crash in the ctest and should make the CI green.

[fixSplinesCTest.patch.tar.gz](https://github.com/AliceO2Group/AliceO2/files/4586232/fixSplinesCTest.patch.tar.gz)

**Comment by matthias-kleiner** [comment]
> @matthias-kleiner : Please add also the attached patch, it fixes the crash in the ctest and should make the CI green.
> 
> [fixSplinesCTest.patch.tar.gz](https://github.com/AliceO2Group/AliceO2/files/4586232/fixSplinesCTest.patch.tar.gz)

Thanks for the fixes. I implemented them

**Pull request merged**